webui/react/src/stores/experiments.tsx

```diff
@@ -1,63 +1,99 @@
+  #experimentMap: WritableObservable<Map<number, ExperimentItem>> = observable(Map());
 
-export const useFetchExperiments = (
-  params: GetExperimentsParams,
-  canceler: AbortController,
-): (() => Promise<void>) => {
-  const context = useContext(ExperimentsContext);
-  if (context === null) {
-    throw new Error('Attempted to use useFetchExperiments outside of Experiment Context');
+  private constructor() {
+    if (!ExperimentsService.#isInternalConstructing) {
+      throw new TypeError('ExperimentsService is not constructable');
+    }
   }
-  const { experimentsCache, updateExperimentsCache } = context;
-
-  return useCallback(async (): Promise<void> => {
-    try {
-      const response = await getExperiments(params, { signal: canceler.signal });
-      updateExperimentsCache(experimentsCache.set(encodeParams(params), response));
-    } catch (e) {
-      handleError(e);
+
+  public static getInstance(): ExperimentsService {
+    if (!ExperimentsService.#instance) {
+      ExperimentsService.#isInternalConstructing = true;
+      ExperimentsService.#instance = new ExperimentsService();
+      ExperimentsService.#isInternalConstructing = false;
     }
-  }, [canceler, experimentsCache, params, updateExperimentsCache]);
-};
+    return ExperimentsService.#instance;
+  }
 
-export const useExperiments = (params: GetExperimentsParams): Loadable<ExperimentPagination> => {
-  const context = useContext(ExperimentsContext);
-  if (context === null) {
-    throw new Error('Attempted to use useExperiments outside of Experiment Context');
+  public experimentsByParams(
+    params: Readonly<GetExperimentsParams>,
+  ): Observable<Loadable<ExperimentPagination>> {
+    return this.#experimentsCache.select((map) => {
+      const cache = map.get(encodeParams(params));
+      if (!cache) {
+        return NotLoaded;
+      }
+      const experiments: ExperimentItem[] = [];
+      const expMap = this.#experimentMap.get();
+      for (const id of cache.experimentIds) {
+        const expItem = expMap.get(id);
+        if (expItem) {
+          experiments.push(expItem);
+        }
+      }
+      const expPagination: ExperimentPagination = {
+        experiments: experiments,
+        pagination: cache.pagination,
+      };
+      return Loaded(expPagination);
+    });
   }
-  const loadedVal = context.experimentsCache.get(encodeParams(params));
 
-  return loadedVal ? Loaded(loadedVal) : NotLoaded;
-};
+  public fetchExperiments(
+    params: Readonly<GetExperimentsParams>,
+    canceler: AbortController,
+  ): () => Promise<void> {
+    return async () => {
+      try {
+        const response = await getExperiments(params, { signal: canceler.signal });
+        this.#updateexperimentsCache(response, params);
+        this.#updateExperimentMap(response);
+      } catch (e) {
+        handleError(e);
+      }
+    };
+  }
+
+  #updateExperimentMap(expPagination: Readonly<ExperimentPagination>) {
```

Comment:
fix: This (and the update cache function below) give the following error during builds:
`SyntaxError: /opt/build/repo/webui/react/src/stores/experiments.tsx: Class private methods are not enabled. Please add `@babel/plugin-proposal-private-methods` to your configuration.`
Either add that plugin to our `craco.config.js` as shown [here](https://stackoverflow.com/questions/68686444/how-to-enable-private-method-syntax-proposal-in-react-app) or remove the `#`.

Reply:
nice catch. compiler didnt catch it while i was working

Reply:
`@babel/plugin-proposal-private-method` help to build app, but couldnt set up for jest. 
i've decided to use ts private for methods, but will replace it with js private after vite migration is done. i think it would be easier to set up `@babel/plugin-proposal-private-method`